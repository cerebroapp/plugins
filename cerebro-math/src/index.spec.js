import { fn } from './index'
import renderer from 'react-test-renderer'

describe('Math plugin', () => {

  it('callback is not called when term is noth math expression', () => {
    const term = 'hello how are you'
    const display = jest.fn()
    fn({ term, display })
    expect(display).not.toBeCalled()
  })

  it('callback called with correct object', () => {
    const term = '5 + 6'
    const display = jest.fn()
    fn({ term, display })
    expect(display).toBeCalledWith({
      icon: 'icon.png',
      clipboard: '11',
      title: '= 11',
      term: '5 + 6 = 11'
    })
  })

  it('formats infinity', () => {
    const term = '1 / 0'
    const display = jest.fn()

    fn({ term, display })
    expect(display).toBeCalledWith({
      icon: 'icon.png',
      clipboard: '∞',
      title: '= ∞',
      term: '1 / 0 = ∞'
    })
  })

  it('shows indeterminate', () => {
    const term = '0 / 0'
    const display = jest.fn()
    fn({ term, display })

    expect(display).toBeCalledWith(expect.objectContaining({
      icon: 'icon.png',
      title: '= indeterminate'
    }))
  })

  it('indeterminate results shows details', () => {
    const term = '0 / 0'
    const display = ({ getPreview }) => {
      const tree = renderer.create(getPreview())
      expect(tree).toMatchSnapshot()
    }
    fn({ term, display })
  })

  describe('with complex expressions: ', () => {
    it('executed several operations', () => {
      const term = '5 * 6 - 10'
      const display = jest.fn()
      fn({ term, display })
      expect(display).toBeCalledWith(expect.objectContaining({
        title: "= 20",
        clipboard: '20'
      }))
    })

    it('uses brackets', () => {
      const term = '5 * (6 - 10)'
      const display = jest.fn()
      fn({ term, display })
      expect(display).toBeCalledWith(expect.objectContaining({
        title: '= -20'
      }))
    })

    it('calculates floats', () => {
      const term = '1.1 + 3.2'
      const display = jest.fn()
      fn({ term, display })
      expect(display).toBeCalledWith(expect.objectContaining({
        title: '= 4.3'
      }))
    })

    it('floats with commas', () => {
      const term = '1,1 + 3,2'
      const display = jest.fn()
      fn({ term, display })
      expect(display).toBeCalledWith(expect.objectContaining({
        title: '= 4.3'
      }))
    })

    it('handles negatives', () => {
      const term = '-10 + 60'
      const display = jest.fn()
      fn({ term, display })
      expect(display).toBeCalledWith(expect.objectContaining({
        title: '= 50'
      }))
    })
  })
})
