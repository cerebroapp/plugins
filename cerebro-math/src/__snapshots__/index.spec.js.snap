// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Math plugin indeterminate results shows details 1`] = `
<div
  className={undefined}
>
  <p>
    The answer is indeterminate. Why? Because anything can be an anwer, see:
  </p>
  <div
    className={undefined}
  >
    <p>
      Lets suppose that anwer is 
      <span
        className={undefined}
      >
        <sup>
          x
        </sup>
        ⁄
        <sub>
          y
        </sub>
      </span>
    </p>
    <p>
      If 
      <span
        className={undefined}
      >
        <sup>
          0
        </sup>
        ⁄
        <sub>
          0
        </sub>
         = 
        <sup>
          x
        </sup>
        ⁄
        <sub>
          y
        </sub>
      </span>
       then 
      <span
        className={undefined}
      >
        x·0 = y·0
      </span>
    </p>
    <p>
      And it is correct, because 
      <span
        className={undefined}
      >
        0 = 0
      </span>
      .
    </p>
    <p>
      So any 
      <span
        className={undefined}
      >
        x
      </span>
       and 
      <span
        className={undefined}
      >
        y
      </span>
       could be an anwer.
    </p>
  </div>
</div>
`;
